### PR TITLE
Chat History microservice for chat data persistence

### PR DESCRIPTION
**Pull request by yogeshmpandey**
## Description

Chat History microservice helps us to connect to a MongoDB Database and save the user chat conversations. This microservice can be used along with chat specific OPEA use cases for data persistence. 

## Issues

List the issue or RFC link this PR is working on. If there is no such link, please mark it as `n/a`.

n/a
## Type of change

List the type of change like below. Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would break existing design and interface)

## Dependencies

List the newly introduced 3rd party dependency if exists.

motor==3.4.0

## Tests

Describe the tests that you ran to verify your changes.

Added test file for the microservice.

### CODE REVIEW

File: tests/test_chathistory_mongo.sh [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+export MONGO_HOST=${ip_address}
+export MONGO_PORT=27017
+export DB_NAME=${DB_NAME:-"Conversations"}
+export COLLECTION_NAME=${COLLECTION_NAME:-"test"}
+
+function build_docker_images() {
+    cd $WORKPATH
+    echo $(pwd)
+    docker run -d -p 27017:27017 --name=mongo mongo:latest
```
Comment (chensuyue) [inline tests/test_chathistory_mongo.sh:18]: `--name=xxx` -> `--name=test-comps-xxx`  
We use this prefix for GenAIComps test containers, to make it separate with GenAIExamples CI test. We have limit HW resource, so we should make sure the name won't be conflict.

File: tests/test_chathistory_mongo.sh [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    docker build --no-cache -t opea/chathistory-mongo-server:latest --build-arg https_proxy=$https_proxy --build-arg http_proxy=$http_proxy -f comps/chathistory/mongo/docker/Dockerfile .
```
Comment (chensuyue) [inline tests/test_chathistory_mongo.sh:20]: `opea/chathistory-mongo-server:latest` -> `opea/chathistory-mongo-server:comps`

File: tests/test_chathistory_mongo.sh [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+}
+
+function start_service() {
+
+    docker run -d --name="chathistory-mongo-server" -p 6013:6013 -p 6012:6012 -p 6014:6014 -e http_proxy=$http_proxy -e https_proxy=$https_proxy -e no_proxy=$no_proxy -e MONGO_HOST=${MONGO_HOST} -e MONGO_PORT=${MONGO_PORT} -e DB_NAME=${DB_NAME} -e COLLECTION_NAME=${COLLECTION_NAME} opea/chathistory-mongo-server:latest
```
Comment (chensuyue) [inline tests/test_chathistory_mongo.sh:25]: --name=xxx -> --name=test-comps-xxx

File: tests/test_chathistory_mongo.sh [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        echo "Result correct."
+    else
+        echo "Result wrong."
+        exit 1
+    fi
+
+}
+
+function stop_docker() {
+    cid=$(docker ps -aq --filter "name=mongo*")
```
Comment (chensuyue) [inline tests/test_chathistory_mongo.sh:51]: "name=test-comps-*"

File: tests/test_chathistory_mongo.sh (hunk 3 above)
Comment (chensuyue) [reply tests/test_chathistory_mongo.sh:25]: There are steps to clean up containers with "test-comps-" prefix in the CI workflow, so even the test failed, the CI can clean up the docker containers and release the port.

File: tests/test_chathistory_mongo.sh (hunk 4 above)
Comment (yogeshmpandey) [reply tests/test_chathistory_mongo.sh]: Fixed

File: tests/test_chathistory_mongo.sh (hunk 3 above)
Comment (yogeshmpandey) [reply tests/test_chathistory_mongo.sh]: Agreed, Done

File: tests/test_chathistory_mongo.sh (hunk 2 above)
Comment (yogeshmpandey) [reply tests/test_chathistory_mongo.sh]: Done

File: tests/test_chathistory_mongo.sh (hunk 1 above)
Comment (yogeshmpandey) [reply tests/test_chathistory_mongo.sh]: Done

**Review by chensuyue — APPROVED** [review]

File: comps/chathistory/mongo/README.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+Chat History microservice helps us to connect with MongoDB Database and save the user chat conversations.
+
+## Setup Environment Variables
+
+```bash
+export http_proxy=${your_http_proxy}
+export https_proxy=${your_http_proxy}
+export MONGO_HOST=${MONGO_HOST}
+export MONGO_HOST=27017
```
Comment (amberjain1) [inline comps/chathistory/mongo/README.md:11]: MONGO_PORT instead of MONGO_HOST?

File: comps/chathistory/mongo/mongo_store.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+            else:
+                return False
+        except BsonError.InvalidId as e:
+            print(e)
+            raise KeyError(e)
+        except Exception as e:
+            print(e)
+            raise Exception(e)
+
+    async def get_all_documents_of_user(self) -> list[dict]:
```
Comment (amberjain1) [inline comps/chathistory/mongo/mongo_store.py:80]: get_all_user_docs could be a better name?

File: comps/chathistory/mongo/mongo_store.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        """
+
+        try:
+            _id = ObjectId(document_id)
+            delete_result = await self.collection.delete_one({"_id": _id, "data.user": self.user})
+
+            delete_count = delete_result.deleted_count
+            print(f"Deleted {delete_count} documents!")
+
+            return True if delete_count == 1 else False
```
Comment (amberjain1) [inline comps/chathistory/mongo/mongo_store.py:148]: Is false a valid flow? or an exception?

Comment (yogeshmpandey) [reply comps/chathistory/mongo/mongo_store.py]: Fixed it to raise an exception.

File: comps/chathistory/mongo/README.md (hunk 5 above)
Comment (yogeshmpandey) [reply comps/chathistory/mongo/README.md]: Fixed

File: comps/chathistory/mongo/mongo_store.py (hunk 6 above)
Comment (yogeshmpandey) [reply comps/chathistory/mongo/mongo_store.py:81]: get_all_user_docs I think translates to get all user's documents which might create confusion.

**Review by ftian1 — COMMENTED** [review]
could you explain how this component works with other microservice like LLM? having a workflow chart/flow in README would be great:)

**Comment by yogeshmpandey** [comment]
> could you explain how this component works with other microservice like LLM? having a workflow chart/flow in README would be great:)

@ftian1 I have updated the README with a flow diagram. The microservice is designed on the same principles as the dataprep component. We plan to use this microservice to add data persistence in the ChatQnA example in upcoming releases.

**Review by amberjain1 — APPROVED** [review]

**Review by ftian1 — APPROVED** [review]

**Pull request merged**